lib/posix-poll: Autoremove closed files from epoll

andreittr opened the pull request:
### Description of changes

This change makes epoll automatically unregister any monitored files when the last reference to them is released, making use of ukfile finalizers.
Furthermore, to enable this change, vfscore vs ukfile shimming in epoll code has also been cleaned up a bit.

This behavior helps mitigate a design oversight of `epoll` where if applications close file descriptors before removing them from epoll's monitored set, they lose the ability to deregister said file without closing the entire epoll fd.

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration
N/A

andreittr committed:
lib/posix-poll: Autoremove closed files from epoll

This change makes epoll automatically unregister any monitored files
when the last reference to them is released, making use of file
finalizers.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

DeliaPavel reviewed (approved):
Looks good to me!
Reviewed-by: Delia Pavel <delia_maria.pavel@stud.acs.upb.ro>

andreistan26 commented on lib/posix-poll/epoll.c:
```diff
@@ -182,16 +199,42 @@ static struct epoll_entry **epoll_find(const struct uk_file *epf, int fd,
+	sf.file = ent->f;
+	uk_file_wlock(ent->epf);
+
+	uk_pr_err("Autodel %p\n", arg);
```
Why is this an error?

andreittr replied:
Oh my, good catch. It's mistakenly left over from development, will fix (remove or turn to info).

andreistan26 reviewed (commented):
Other than that it works great!

andreittr commented:
PR ON HOLD pending fixing an unforeseen bug re: autoremoving lwip sockets. Turned to draft.

andreittr committed:
lib/ukfile: Ensure finalizers run after destructor

This change ensures that file finalizers are executed after the main
file destructor when the last strong reference to a file is released.
Finalizers may themselves release weak references, which in turn may
trigger the file destructor. Previously this could lead to destructors
being called in the wrong order.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr commented:
PR reopened; bug fixed & comments addressed. Reviews re-requested.

andreistan26 reviewed (approved):
Looks good!

Reviewed-by: Andrei Stan [andreistan2003@gmail.com](mailto:andreistan2003@gmail.com)

DeliaPavel reviewed (approved):
Reviewed-by: Delia Pavel <delia_maria.pavel@stud.acs.upb.ro>

eduardvintila reviewed (approved):
All good, nice job.

Reviewed-by: Eduard Vintilă <eduard.vintila47@gmail.com>

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
